DSharpPlus/Clients/DiscordClient.cs

```diff
@@ -136,12 +136,13 @@ public int Ping
         public DiscordClient(DiscordConfiguration config)
             : base(config)
         {
-            if (this.Configuration.MessageCacheSize > 0)
+            var intents = this.Configuration.Intents;
+            bool isEnableCache = (intents.HasIntent(DiscordIntents.GuildMessages) || intents.HasIntent(DiscordIntents.DirectMessages))
+                && ((this.Configuration.MessageCacheProvider == null && this.Configuration.MessageCacheSize > 0) || this.Configuration.MessageCacheProvider != null);
```

Comment:
Surely this could be simplified...

Reply:
Also inline the variable

Reply:
It's tested with a small bot that re-sends a message when it's deleted from Discord, which should cover all three methods provided by the cache. On the custom implementation part, it's just tested to make sure that method calls are wired properly to it, lol. 

Also I slightly cleaned up this part, should look better now.